File: benchmarks/simulation/schedule_strategies/fcfs.py
from typing import Dict, List, Optional, Tuple
from .base import ScheduleStrategy, register_schedule_strategy
from ..simulator import RequestStats
from ..dependency_graph import GraphNode, AttnNode, ExpertNode
from ..logger import logger
from vllm.transformers_utils.cost_model import ModelComponent

@register_schedule_strategy('FCFS')
class FCFS(ScheduleStrategy):
    def __init__(self,
                 n_layers: int,
                 per_token_latency_slo_ms: float
                 ) -> None:
        self.n_layers: int = n_layers
        self.per_token_latency_slo_ms: float = per_token_latency_slo_ms
        self._current_batch_requests: List[Tuple(int, int)] = []
        self._prev_batch_requests: List[Tuple(int, int)] = []
        self._current_layer: int = 0
        self._current_phase: type[GraphNode] = AttnNode
        self._activated_experts_per_batch = 0
        self._activated_experts_history = []

    def _advance_phase(self) -> None:
        if self._current_phase == AttnNode:
            self._current_phase = ExpertNode
        else:
            self._current_phase = AttnNode
            self._current_layer += 1
        logger.debug("Advanced phase to {}, layer {}.".format(
            self._current_phase.__name__, 
            self._current_layer)
        )

    def _reset_batch(self) -> None:
        self._prev_batch_requests = self._current_batch_requests
        self._current_batch_requests = []
        self._current_layer = 0
        self._current_phase = AttnNode
        self._activated_experts_history.append(self._activated_experts_per_batch)
        self._activated_experts_per_batch = 0

    def schedule(self,
                 stats_dict: Dict[int, RequestStats],
                 ready_nodes: List[GraphNode],
                 max_batch_size: Optional[int] = None,
                 ) -> Tuple[ModelComponent, List[GraphNode]]:
        if not ready_nodes:
            # no more requests to schedule
            return None, None
        # execute the same requests layer by layer
        if not self._current_batch_requests:
            assert self._current_layer == 0
            # first try to execute the next token of the current batch
            ready_first_layer_attn_nodes = []
            for node in ready_nodes:
                if (node.req_id, node.token_index - 1) in self._prev_batch_requests:
                    ready_first_layer_attn_nodes.append(node)
                    self._current_batch_requests.append((node.req_id, node.token_index))
            if len(ready_first_layer_attn_nodes) < max_batch_size:
                # schedule all first layer Attn requests in FCFS order
                for node in sorted(ready_nodes, key=lambda node: stats_dict[node.req_id].enqueue_time):
                    if node.layer_id == self._current_layer and isinstance(node, AttnNode):
                        if (node.req_id, node.token_index) not in self._current_batch_requests:
                            ready_first_layer_attn_nodes.append(node)
                            self._current_batch_requests.append((node.req_id, node.token_index))
                            if len(ready_first_layer_attn_nodes) == max_batch_size:
                                break
            # truncate to max batch size
            logger.debug("Scheduling first batch of {} requests.".format(len(self._current_batch_requests)))
            assert len(self._current_batch_requests) > 0
        # first test if the current batch is all finished
        # we assume all nodes are DecodeNodes
        current_batch_ready_nodes = [node for node in ready_nodes
                                        if (node.req_id, node.token_index) in self._current_batch_requests]
        if not current_batch_ready_nodes:
            # # DEBUG
            # logger.debug("Current batch finished, reset batch")
            # return None, None
            self._reset_batch()
            return self.schedule(stats_dict, ready_nodes, max_batch_size)
        # execute the current batch
        current_batch_ready_nodes = [node for node in current_batch_ready_nodes if
                                     isinstance(node, self._current_phase)
                                     and node.layer_id == self._current_layer]
        if not current_batch_ready_nodes:
            # advance to the next phase
            self._advance_phase()
            current_batch_ready_nodes = [node for node in ready_nodes if
                                     isinstance(node, self._current_phase)
                                     and node.layer_id == self._current_layer]
        if self._current_phase == ExpertNode:
            # schedule one expert at a time
            current_batch_ready_nodes: List[ExpertNode]
            current_batch_ready_nodes = [node for node in current_batch_ready_nodes
                                         if node.expert_id == current_batch_ready_nodes[0].expert_id]
            logger.debug("Scheduling {} expert nodes: Layer {}, Expert: {}".format(
                len(current_batch_ready_nodes),
                self._current_layer,
                current_batch_ready_nodes[0].expert_id
            ))
            self._activated_experts_per_batch += 1
            return ModelComponent.EXPERT, current_batch_ready_nodes
        else:
            # schedule all attn nodes together
            logger.debug("Scheduling {} attn nodes: Layer {}".format(len(current_batch_ready_nodes), self._current_layer))
            return ModelComponent.ATTENTION_GATE, current_batch_ready_nodes

File: benchmarks/simulation_cpp/schedule_strategies/fcfs.cpp
#include "fcfs.hpp"

#include <algorithm>

namespace stragegies {
namespace fcfs_strategy {

using dependency_graph::AttnNode;
using dependency_graph::AttnNodes;
using dependency_graph::ExpertNode;
using dependency_graph::ExpertNodes;
using dependency_graph::GraphNode;
using dependency_graph::GraphNodePtr;
using dependency_graph::GraphNodes;

FCFSStrategy::FCFSStrategy(const RequestGraphs& graphs,
                           const StrategyConfig& config)
    : Strategy(graphs, config) {
  n_layers_ = config.GetInt("n_layers");
  max_batch_size_ = config.GetInt("max_batch_size");
  max_kv_tokens_ = config.GetInt("max_kv_tokens");
  per_token_latency_slo_ = config_.GetFloat("per_token_latency_slo");
}

ScheduleResult FCFSStrategy::Schedule(
    const std::unordered_map<int, RequestStats>& request_stats,
    const GraphNodes& ready_nodes, float current_time) {
  if (ready_nodes.empty()) {
    // no more requests to schedule
    return ScheduleResult(ModelComponent::kAttnGate, {});
  }
  // execute the same requests layer by layer
  if (current_batch_requests_.empty()) {
    if (current_layer_ != 0) {
      throw std::runtime_error(
          "current_batch_requests_ is empty but current_layer_ is not 0");
    }
    // first try to execute the next token of the current batch
    int current_batch_kv_tokens = 0;
    GraphNodes ready_first_layer_attn_nodes;
    for (const auto& node : ready_nodes) {
      if (node->layer_id == 0 && node->Type() == NodeType::kAttn) {
        ready_first_layer_attn_nodes.push_back(node);
      }
    }
    for (const auto& node : ready_first_layer_attn_nodes) {
      if (prev_batch_requests_.find({node->req_id, node->token_index - 1}) !=
          prev_batch_requests_.end()) {
        current_batch_requests_.insert({node->req_id, node->token_index});
        current_batch_kv_tokens +=
            request_stats.at(node->req_id).prompt_len + request_stats.at(node->req_id).GetNumTokensDecoded();
      }
    }
    if ((int)current_batch_requests_.size() < max_batch_size_ &&
        current_batch_kv_tokens < max_kv_tokens_) {
      if (current_layer_ != 0 || current_phase_ != NodeType::kAttn) {
        throw std::runtime_error(
            "ready_first_layer_attn_nodes is not full but current_layer_ is "
            "not 0 or current_phase_ is not kAttn");
      }
      // schedule all first layer Attn requests in FCFS order
      // first sort ready nodes by enqueue time
      GraphNodes sorted_ready_nodes = ready_first_layer_attn_nodes;
      std::sort(sorted_ready_nodes.begin(), sorted_ready_nodes.end(),
                [&](const GraphNodePtr a, const GraphNodePtr b) {
                  return request_stats.at(a->req_id).enqueue_time <
                         request_stats.at(b->req_id).enqueue_time;
                });
      for (const auto& node : sorted_ready_nodes) {
        if (current_batch_requests_.find({node->req_id, node->token_index}) ==
            current_batch_requests_.end()) {
          current_batch_requests_.insert({node->req_id, node->token_index});
          current_batch_kv_tokens +=
              request_stats.at(node->req_id).GetTotalContextLength();
          if ((int)current_batch_requests_.size() == max_batch_size_ ||
              current_batch_kv_tokens >= max_kv_tokens_) {
            break;
          }
        }
      }
    }
    LOG("Scheduling a batch of "
        << current_batch_requests_.size()
        << " requests, total KV size: " << current_batch_kv_tokens << " / "
        << max_kv_tokens_ << " tokens." << std::endl);
    UpdateAvgBatchSize_(current_batch_requests_.size());
  }
  // first test if the current batch is all finished
  // we assume all nodes are DecodeNodes
  GraphNodes current_batch_ready_nodes;
  for (const auto& node : ready_nodes) {
    if (current_batch_requests_.find({node->req_id, node->token_index}) !=
        current_batch_requests_.end()) {
      current_batch_ready_nodes.push_back(node);
    }
  }
  if (current_batch_ready_nodes.empty()) {
    // current batch is all finished
    // advance to the next batch
    ResetBatch_();
    return Schedule(request_stats, ready_nodes, current_time);
  }
  // filter out nodes that are not in the current phase
  GraphNodes current_phase_ready_nodes;
  for (const auto& node : current_batch_ready_nodes) {
    if (node->Type() == current_phase_ && node->layer_id == current_layer_) {
      current_phase_ready_nodes.push_back(node);
    }
  }
  if (current_phase_ready_nodes.empty()) {
    // current phase is all finished
    // advance to the next phase
    AdvancePhase_();
    current_phase_ready_nodes.clear();
    for (const auto& node : current_batch_ready_nodes) {
      if (node->Type() == current_phase_ && node->layer_id == current_layer_) {
        current_phase_ready_nodes.push_back(node);
      }
    }
  }
  // schedule the current phase
  if (current_phase_ == NodeType::kExpert) {
    // schedule one expert at a time
    ExpertNodes ready_expert_nodes;
    for (const auto& node : current_phase_ready_nodes) {
      ready_expert_nodes.push_back(
          std::dynamic_pointer_cast<ExpertNode>(node));
    }
    if (ready_expert_nodes.empty()) {
      throw std::runtime_error(
          "current_phase_ready_nodes is empty but current_phase_ is "
          "kExpert");
    }
    int expert_id_to_schedule = ready_expert_nodes[0]->expert_id;
    GraphNodes nodes_to_schedule;
    for (const auto& node : ready_expert_nodes) {
      if (node->expert_id == expert_id_to_schedule) {
        nodes_to_schedule.push_back(node);
      }
    }
    activated_experts_per_layer_ += 1;
    return {ModelComponent::kExpert, nodes_to_schedule};
  }
  // schedule all attn nodes together
  return {ModelComponent::kAttnGate, current_phase_ready_nodes};
}

void FCFSStrategy::AdvancePhase_() {
  if (current_phase_ == NodeType::kAttn) {
    current_phase_ = NodeType::kExpert;
  } else {
    UpdateAvgActivatedExperts_(activated_experts_per_layer_);
    activated_experts_per_layer_ = 0;
    current_phase_ = NodeType::kAttn;
    current_layer_ += 1;
  }
}

void FCFSStrategy::ResetBatch_() {
  UpdateAvgActivatedExperts_(activated_experts_per_layer_);
  activated_experts_per_layer_ = 0;
  prev_batch_requests_ = current_batch_requests_;
  current_batch_requests_.clear();
  current_phase_ = NodeType::kAttn;
  current_layer_ = 0;
}

float FCFSStrategy::GetAvgActivatedExperts() const {
  return avg_activated_experts_.first;
}

float FCFSStrategy::GetAvgBatchSize() const { return avg_batch_size_.first; }

void FCFSStrategy::UpdateAvgActivatedExperts_(int n_experts) {
  avg_activated_experts_.first =
      (avg_activated_experts_.first * avg_activated_experts_.second +
       n_experts) /
      (avg_activated_experts_.second + 1);
  avg_activated_experts_.second += 1;
}

void FCFSStrategy::UpdateAvgBatchSize_(int batch_size) {
  avg_batch_size_.first =
      (avg_batch_size_.first * avg_batch_size_.second + batch_size) /
      (avg_batch_size_.second + 1);
  avg_batch_size_.second += 1;
}

}  // namespace fcfs_strategy
}  // namespace stragegies